feat(infra): add dual-mode support for Granian and Unit servers

webjunkie opened the pull request:
## Summary
Add dual-mode Docker image supporting both Granian (default) and Nginx Unit servers via `USE_GRANIAN` environment variable.

## Motivation
- Test Granian in all environments by default for maximum coverage
- Provide safety valve for critical production pods to stay on Unit
- Enable gradual rollout without building separate images

## Changes
- Add `granian[uvloop,reload,pname]==2.5.5` alongside existing gunicorn in pyproject.toml
- Add `bin/granian_metrics.py` for Prometheus metrics aggregation in Granian mode
- Modify `bin/docker-server-unit` to support dual mode via `USE_GRANIAN` env var:
  - `USE_GRANIAN=true` (default): Runs Granian ASGI server with 4 workers, 2 threads, uvloop
  - `USE_GRANIAN=false`: Runs Nginx Unit server (existing behavior)
- Update `bin/start-backend` to use Granian with hot reload for local development
- Both modes use `PROMETHEUS_MULTIPROC_DIR` for multi-process metrics aggregation on port 8001

## Configuration

**Default (Granian):**
```bash
# No env var needed - Granian runs by default
docker run posthog/posthog
```

**Fallback (Unit):**
```bash
# For critical production pods during gradual rollout
docker run -e USE_GRANIAN=false posthog/posthog
```

## Performance
Granian provides 3-12x performance improvement over uvicorn with better HTTP/2 and WebSocket support.

## Image Size
No increase - uses existing `unit:1.33.0-python3.12` base image (~387MB) which already has Python 3.12.

## Rollout Strategy
1. Default: Granian gets tested everywhere (staging, dev, new deployments)
2. Safety valve: Critical production pods can set `USE_GRANIAN=false` 
3. Gradual migration: Move production pods to Granian as confidence builds
4. Future: Remove Unit support after successful rollout

## Testing
- Local dev already using Granian via `bin/start-backend`
- Metrics endpoint (port 8001) works in both modes
- Both servers tested on `unit:1.33.0-python3.12` base image

## Dashboards
Note: Granian exposes basic config metrics (`granian_workers_total`, `granian_max_worker_threads`) plus Django app metrics. Unit-specific metrics (`unit_application_*`) won't be available in Granian mode. New Granian dashboard will be needed in posthog-cloud-infra.

webjunkie committed:
feat(infra): add dual-mode support for Granian and Unit servers

Add ability to switch between Granian (default) and Nginx Unit via USE_GRANIAN environment variable.

Changes:
- Add granian[uvloop,reload,pname]==2.5.5 alongside existing gunicorn
- Add bin/granian_metrics.py for Prometheus metrics on port 8001 in Granian mode
- Modify bin/docker-server-unit to support dual mode via USE_GRANIAN env var
  - USE_GRANIAN=true (default): Runs Granian ASGI server with 4 workers, 2 threads, uvloop
  - USE_GRANIAN=false: Runs Nginx Unit server (existing behavior)
- Update bin/start-backend to use Granian with hot reload for local development
- Both modes use PROMETHEUS_MULTIPROC_DIR for multi-process metrics aggregation

Default is Granian for better performance (3-12x faster than uvicorn, better HTTP/2 and WebSocket support).
Unit remains available as fallback via USE_GRANIAN=false for safe rollback if needed.

Dockerfile uses existing unit:1.33.0-python3.12 base which already has Python 3.12, so no image changes needed.

webjunkie committed:
fix(infra): revert to Granian as default for broader testing

Change default back to Granian so it gets tested everywhere by default.
Production pods that need Unit stability can explicitly set USE_GRANIAN=false.

- Default: Granian (USE_GRANIAN=true or unset) - gets tested in all environments
- Opt-out: Unit (USE_GRANIAN=false) - for critical production pods during gradual rollout
- Local dev: Granian via bin/start-backend

This ensures Granian gets maximum testing coverage while giving production safety valve.

webjunkie committed:
chore(infra): remove gunicorn - unused in dual-mode setup

Remove gunicorn and related files as neither Granian nor Unit modes use it:
- Remove gunicorn==20.1.0 from pyproject.toml
- Remove gunicorn.config.py file and Dockerfile reference
- Remove bin/docker-server (gets replaced by symlink to docker-server-unit at build time)
- Update comments to reference granian_metrics.py and unit_metrics.py instead

Both server modes now clearly defined:
- Granian mode: Uses granian with bin/granian_metrics.py for metrics
- Unit mode: Uses Nginx Unit with bin/unit_metrics.py for metrics

chore(infra): remove gunicorn - unused in dual-mode setup

Remove gunicorn and related files as neither Granian nor Unit modes use it:
- Remove gunicorn==20.1.0 from pyproject.toml
- Remove gunicorn.config.py file and Dockerfile reference
- Remove bin/docker-server (gets replaced by symlink to docker-server-unit at build time)
- Update comments to reference granian_metrics.py and unit_metrics.py instead

Both server modes now clearly defined:
- Granian mode: Uses granian with bin/granian_metrics.py for metrics
- Unit mode: Uses Nginx Unit with bin/unit_metrics.py for metrics

webjunkie committed:
chore(infra): merge latest master with dual-mode changes

Resolve conflicts between dual-mode Granian/Unit implementation and master updates:
- Use master's updated dependency versions (google-genai 1.46.0, hogql-parser 1.2.11)
- Keep granian[uvloop,reload,pname]==2.5.5 from dual-mode branch
- Regenerate uv.lock with resolved dependencies

webjunkie committed:
fix(infra): restore docker-server wrapper for backward compatibility

Restore bin/docker-server as a simple wrapper that delegates to docker-server-unit.
This fixes hobby deployments and local Docker scripts that call ./bin/docker-server.

The wrapper approach is cleaner than the Dockerfile symlink because:
- Works for both local scripts and Docker container
- Maintains git history and permissions
- Simpler to understand and maintain

webjunkie committed:
fix(infra): address code review feedback

- Combine trap handlers in docker-server-unit to prevent temp dir cleanup failure
  (second trap was overwriting first, preventing PROMETHEUS_MULTIPROC_DIR cleanup)
- Remove test comments from posthog/asgi.py

webjunkie committed:
chore(infra): make Unit the default for safer production rollout

Change default from Granian to Unit for lower risk during gradual rollout:
- Production: Unit by default (proven stable, less risk)
- Opt-in to Granian: Set USE_GRANIAN=true
- Local dev: Always uses Granian (bin/start-backend) for better DX

This allows conservative rollout where we can test Granian in specific pods
before switching the default.

langesven reviewed (approved):
definitely agree with testing it again 😄

but this looks good I think, I mean right now nothing except for local dev should change when this is merged. then we can toggle on granian mode for dev workloads

webjunkie committed:
fix(infra): add cleanup trap for Unit path

Add missing trap to cleanup PROMETHEUS_MULTIPROC_DIR in the Unit code path.
Both Granian and Unit paths now properly cleanup the temp directory on exit.

webjunkie committed:
chore: merge master into dual-mode branch

Resolve conflict in hogli manifest - keep both granian-metrics and sync-storage tools
